vm-allocator/src/system.rs

```diff
@@ -0,0 +1,152 @@
+///    assert_eq!(allocator.allocate_irq(), Some(6));
+///    assert_eq!(allocator.allocate_mmio_addresses(None, 0x1000), Some(GuestAddress(0x1fffe000)));
+///
+/// ```
+pub struct SystemAllocator {
+    io_address_space: Option<AddressAllocator>,
+    mmio_address_space: AddressAllocator,
+    irq: IdAllocator,
+    instance_id: IdAllocator,
+}
+
+impl SystemAllocator {
+    /// Creates a new `SystemAllocator` for managing addresses and irq numvers.
+    /// Can return `None` if `base` + `size` overflows a u64 or if alignment isn't a power
+    /// of two.
+    ///
+    /// * `io_base` - The starting address of IO memory.
+    /// * `io_size` - The size of IO memory.
+    /// * `mmio_base` - The starting address of MMIO memory.
+    /// * `mmio_size` - The size of MMIO memory.
+    /// * `first_irq` - The first irq number to give out.
+    /// * `last_irq` - The last irq number to give out.
+    /// * `first_instance_id` - The first device instance id to give out.
+    pub fn new(
+        io_base: Option<GuestAddress>,
+        io_size: Option<GuestUsize>,
+        mmio_base: GuestAddress,
+        mmio_size: GuestUsize,
+        first_irq: u32,
+        last_irq: u32,
+        first_instance_id: u32,
+    ) -> Option<Self> {
+        let page_size = pagesize() as u64;
+        Some(SystemAllocator {
+            io_address_space: if let (Some(b), Some(s)) = (io_base, io_size) {
+                Some(AddressAllocator::new(b, s, Some(0x1))?)
+            } else {
+                None
+            },
+            mmio_address_space: AddressAllocator::new(mmio_base, mmio_size, Some(page_size))?,
+            irq: IdAllocator::new(first_irq, last_irq)?,
+            instance_id: IdAllocator::new(first_instance_id, u32::max_value())?,
+        })
+    }
+
+    /// Reserves the next available system irq number.
+    /// * `irq` - A specific value trying to allocate, or None means no specific value.
+    pub fn allocate_irq(&mut self, irq: Option<u32>) -> Result<u32> {
+        self.irq.allocate(irq).map_err(Error::IdAllocate)
+    }
+
+    /// Reserves the next available system device instance id number.
+    pub fn allocate_instance_id(&mut self) -> Result<u32> {
+        self.instance_id.allocate(None).map_err(Error::IdAllocate)
+    }
+
+    /// Free an interrupt number.
+    /// Only free an `irq` if it matches exactly an already allocated one.
+    pub fn free_irq(&mut self, irq: Option<u32>) {
+        match irq {
+            Some(i) => self.irq.free(i),
+            None => return,
+        }
+    }
+
+    /// Free an instance id.
+    /// Only free an `id` if it matches exactly an already allocated one.
+    pub fn free_instance_id(&mut self, id: u32) {
+        self.instance_id.free(id);
+    }
+
+    /// Reserves a section of `size` bytes of IO address space.
+    pub fn allocate_io_addresses(
+        &mut self,
+        address: GuestAddress,
+        size: GuestUsize,
+    ) -> Option<GuestAddress> {
```

Comment:
Part of my previous comments that may not have been clear enough is that we want the APIs to return `Result` with the appropriate error instead of `Option`. So both `system.rs` and `address.rs` should return a `Result`.

Reply:
Also, another part of the comments were that our `system.rs` API should be thread safe, and that we should associate a Mutex with each individual allocator that's part of `SystemAllocator`. Have you addressed that comment?

Reply:
@sameo  Let me add the protection in system.rs. Updated.